C7GameData/Tile.cs

```diff
@@ -55,12 +58,148 @@ public class Tile
 
 		public TileOverlays overlays = new TileOverlays();
 
+
+		// -1 = not initialized. 0 = false, 1 = true
+		public int isLake = -1;
+
 		public Tile()
 		{
 			unitsOnTile = new List<MapUnit>();
 			Resource = Resource.NONE;
 		}
 
+		public bool IsLake()
+		{
+			// If the isLake value has been initialized, we should return that
+			if(isLake != -1)
+			{
+				return (isLake == 1);
+			}
+
+			if(!IsWater())
+			{
+				isLake = 0;
+				return false;
+			}
+
+			int contiguousWater = 0;
+			List<Tile> contiguous = new List<Tile>();
+			List<Tile> searchedWater = new List<Tile>();
+			contiguous.Add(this);
+
+			// Default is that this *is* a lake because other conditions will assure we only get water tiles. If the lake is under size 21, we want to default to it being a lake
+			bool lake = true;
+
+			int i = 0;
+			while(i < contiguous.Count())
+			{
+				Tile currentTile = contiguous[i];
+
+				// Skip land tiles and remove from the stack. This should be checked before they're added but just in case
+				if(!currentTile.IsWater())
+				{
+					currentTile.isLake = 0;
+					contiguous.Remove(currentTile);
+					continue;
+				}
+				contiguousWater += 1;
+				if(contiguousWater > 20)
```

Comment:
Also minor, but this long block of if/else/foreach is a good example of why I recommend the style:

```
if (something) {
   //do something
}
```

rather than:

```
if (something)
{
    //do something
}
```

When there are a lot of checks like this, the extra lines from that leading { add up to making it harder to read the logic at a glance due to having to scroll more to see the whole picture, with each scroll adding a little bit of time and overhead re-focusing the eye on where the most-recently-read line is now.  Not a big deal and I see we've been inconsistent about it, but that's the reason I eventually switched to that style from originally giving the opening curly brace its own line.

Reply:
One last style comment, I realized that `i` and `contiguousWater` are both counting the same thing - how many water tiles we have.  They're just zero-indexed to account for the array and one-indexed because that's how humans think of counting things.  We want to say, if we find a 21st water tile, it's not a lake.  

I'd propose just having one counter variable, something like waterTileCount.  Start it at 0 and increment it in the same place `contiguousWater` is incremented, and it will both grab the right tile from the array and read nicely.  That way someone reading this code won't be trying to figure out the difference between two counters and why one is incremented early and one is incremented late - it'll be easier to verify that the logic works correctly.